File: app/src/main/java/com/github/fo2rist/mclaren/web/StoryStreamWebServiceImpl.java
package com.github.fo2rist.mclaren.web;

import android.support.annotation.NonNull;

import com.github.fo2rist.mclaren.BuildConfig;
import javax.inject.Inject;
import javax.inject.Named;
import javax.inject.Singleton;
import okhttp3.Headers;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;

@Singleton
public class StoryStreamWebServiceImpl implements StoryStreamWebService {

    @NonNull
    private static final HttpUrl FEED_URL = HttpUrl.parse(BuildConfig.STORYSTREAM_FEED_URL);
    private static final String ACCESS_TOKEN = BuildConfig.STORYSTREAM_TOKEN;
    private static final int STORIES_PER_PAGE = 20;
    private static final boolean INCLUDE_ALL_MEDIA = true;
    private static final Headers DEFAULT_HEADERS = Headers.of(
            "Content-Type", "application/json; charset=utf-8"
    );

    private final OkHttpClient client;

    @Inject
    StoryStreamWebServiceImpl(@Named("web-okhttp") OkHttpClient webClient) {
        this.client = webClient;
    }

    @Override
    public void requestLatestFeed(@NonNull FeedRequestCallback callback) {
        client.newCall(createLatestFeedRequest())
                .enqueue(new FeedCallbackWrapper(FeedWebService.DEFAULT_PAGE, callback));
    }

    @Override
    public void requestFeedPage(int pageNumber, @NonNull FeedRequestCallback callback) {
        client.newCall(createFeedPageRequest(pageNumber))
                .enqueue(new FeedCallbackWrapper(pageNumber, callback));
    }

    private Request createLatestFeedRequest() {
        return createFeedPageRequest(1); //StoryStream reacts on page=1 properly so we can use it for the top page
    }

    private Request createFeedPageRequest(int pageNumber) {
        HttpUrl url = FEED_URL
                .newBuilder()
                .addQueryParameter("access_token", ACCESS_TOKEN)
                .addQueryParameter("page", String.valueOf(pageNumber))
                .addQueryParameter("rpp", String.valueOf(STORIES_PER_PAGE))
                .addQueryParameter("all_media", String.valueOf(INCLUDE_ALL_MEDIA))
                .build();

        return new Request.Builder()
                .url(url)
                .headers(DEFAULT_HEADERS)
                .build();
    }
}

File: app/src/main/java/com/github/fo2rist/mclaren/web/StoryStreamWebServiceImpl.kt
package com.github.fo2rist.mclaren.web

import android.support.annotation.VisibleForTesting
import com.github.fo2rist.mclaren.BuildConfig
import com.github.fo2rist.mclaren.web.utils.executeAsync
import okhttp3.Headers
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Singleton
internal class StoryStreamWebServiceImpl @Inject internal constructor(
    @param:Named("web-okhttp")
    private val client: OkHttpClient
) : StoryStreamWebService {

    override suspend fun requestLatestFeed(): String? {
        return client.newCall(createLatestFeedRequest()).executeAsync()
    }

    override suspend fun requestFeedPage(pageNumber: Int): String? {
        return client.newCall(createFeedPageRequest(pageNumber)).executeAsync()
    }

    private fun createLatestFeedRequest(): Request = createFeedPageRequest()

    @VisibleForTesting
    fun createFeedPageRequest(pageNumber: Int = 1): Request {
        // StoryStream reacts on page=1 properly so we can use #1 for the top page.
        val url = FEED_URL
                .newBuilder()
                .addQueryParameter("access_token", ACCESS_TOKEN)
                .addQueryParameter("page", pageNumber.toString())
                .addQueryParameter("rpp", STORIES_PER_PAGE.toString())
                .addQueryParameter("all_media", INCLUDE_ALL_MEDIA.toString())
                .build()

        return Request.Builder()
                .url(url)
                .headers(DEFAULT_HEADERS)
                .build()
    }

    companion object {

        private val FEED_URL = HttpUrl.get(BuildConfig.STORYSTREAM_FEED_URL)
        private const val ACCESS_TOKEN = BuildConfig.STORYSTREAM_TOKEN
        private const val STORIES_PER_PAGE = 20
        private const val INCLUDE_ALL_MEDIA = true
        private val DEFAULT_HEADERS = Headers.of(
                "Content-Type", "application/json; charset=utf-8"
        )
    }
}
